head/src/main/kotlin/io/qalipsis/core/head/report/thymeleaf/ThymeleafReportServiceImpl.kt

```diff
@@ -0,0 +1,174 @@
+            addDialect(Java8TimeDialect())
+        }
+        Files.createDirectories(Paths.get(CHART_IMAGE_DIRECTORY))
+        val chartImages = campaignReportDetail.chartData.mapIndexed { index, chartData ->
+            chartService.buildChart(
+                chartData,
+                dataSeries,
+                index
+            )
+        }
+        val context = buildContext(campaignReportDetail, chartImages)
+
+        return templateEngine.process("report_template", context)
+    }
+
+    /**
+     * Generates a pdf template from HTML string.
+     *
+     * @param html HTML template to be rendered
+     * @param reportName title of the report
+     * @param result file content to be returned wrapped in a Slot
+     */
+    private suspend fun generatePdfFromHtml(html: String, reportName: String, result: Slot<Result<ByteArray>>) {
+        try {
+            require(html.isNotEmpty()) { "HTML File is empty" }
+            val reportFile = File.createTempFile(reportName, ".pdf")
+            reportFile.mkdirs()
+            val outputStream: OutputStream = FileOutputStream(reportFile)
+            ITextRenderer().apply {
+                setDocumentFromString(html)
+                layout()
+                createPDF(outputStream)
+            }
+            outputStream.close()
+            val fileContent = reportFile.readBytes()
+            assert(fileContent.isNotEmpty()) { "File content could not be read" }
+            result.set(Result.success(fileContent))
+        } catch (e: Exception) {
+            result.set(Result.failure(exception = e))
+        }
+    }
+
+    override suspend fun generatePdf(
+        report: ReportEntity,
+        campaignReportDetail: CampaignReportDetail,
+        reportTask: ReportTaskEntity,
+        creator: String,
+        dataSeries: Collection<DataSeriesEntity>,
+        tenant: String
+    ): ByteArray {
+        reportTaskRepository.update(
+            ReportTaskEntity(
+                id = reportTask.id,
+                tenantReference = tenant,
+                reportId = report.id,
+                status = ReportTaskStatus.PROCESSING,
+                creationTimestamp = reportTask.creationTimestamp,
+                updateTimestamp = Instant.now(),
+                creator = creator
+            )
+        )
+        val result = Slot<Result<ByteArray>>()
+        val html = parseThymeleafTemplate(campaignReportDetail, dataSeries)
+        generatePdfFromHtml(html, report.displayName, result)
+
+        return result.get().getOrThrow()
+    }
+
+    /**
+     * Builds up the thymeleaf context.
+     *
+     * @param campaignReportDetail contains data with which to set up the context
+     * @param chartImages name of files containing report chart images
+     */
+    private fun buildContext(campaignReportDetail: CampaignReportDetail, chartImages: List<String>) = Context().apply {
+        setVariable("title", campaignReportDetail.reportName)
```

Comment:
Please explain each step of configuration to understand what is configured there.